### PR TITLE
News and video Site Maps

### PR DESCRIPTION
**Pull request by rich-nguyen**
This is a migration of the old R2 news and video sitemap. Behaviour is roughly the same, except:
- the news site map contains up to 1000 articles from the last two days, rather than the 100 newest articles.
- the video site map was empty, so has similar logic to the new news sitemap.

### CODE REVIEW

File: applications/app/jobs/SitemapLifecycle.scala [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    }
+  }
+}
+
+object SiteMapJob extends ExecutionContexts with Logging {
+  case class SiteMapContent(
+    news: xml.NodeSeq,
+    video: xml.NodeSeq)
+
+  private val siteMapContent = AkkaAgent[SiteMapContent](SiteMapContent(Nil, Nil))
```
Comment (johnduffell) [inline applications/app/jobs/SitemapLifecycle.scala]: is it better to give an empty sitemap before the job completes successfully or an error?

File: applications/app/services/NewsSiteMap.scala [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  private case class Url(
+    location: String,
+    keywords: String,
+    title: String,
+    genres: String,
+    webPublicationDate: DateTime,
+    imageUrl: String) {
+
+    def xml() = {
+      <url>
```
Comment (johnduffell) [inline applications/app/services/NewsSiteMap.scala]: this is really cool!

File: common/app/common/configuration.scala [hunk 3]
```diff
@@ -195,6 +195,7 @@ class GuardianConfiguration(val application: String, val webappConfDirectory: St
 
   object images {
     lazy val path = configuration.getMandatoryStringProperty("images.path")
+    val fallbackLogo = Static("images/fallback-logo.png").path
```
Comment (jennysivapalan) [inline common/app/common/configuration.scala]: Just worth noting the size of fallback-logo.png is currently optimised for facebook.

**Comment by jennysivapalan** [comment]
:+1:

**Comment by jennysivapalan** [comment]
@rich-nguyen can we get this updated and merged?

**Comment by rich-nguyen** [comment]
@jennysivapalan yes! on it. needs router(s) changes of course too.

File: applications/app/services/NewsSiteMap.scala [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        val keywords = (keywordTags ++ sectionTag).mkString(", ")
+
+        val genres = item.tones.flatMap(_.webTitle match {
+          case "Blogposts" => Some("Blog")
+          case "Reviews" => Some("Opinion")
+          case "Comment" => Some("OpEd")
+          case _ => None
+        }).mkString(", ")
+
+        val imageUrl: String = item.mainPicture.flatMap(_.largestEditorialCrop.flatMap(_.url))
```
Comment (JustinPinner) [inline applications/app/services/NewsSiteMap.scala]: Is `largestEditorialCrop` always the largest image available? I think R2 used to select the next image size down. Not that it's necessarily an issue, just a difference.

File: applications/app/services/NewsSiteMap.scala [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        Url(
+          location = item.webUrl,
+          keywords = keywords,
+          title = item.headline,
+          genres = genres,
+          webPublicationDate = item.webPublicationDate,
+          imageUrl = imageUrl)
+      }
+
+      UrlSet(urls take 1000).xml()
```
Comment (JustinPinner) [inline applications/app/services/NewsSiteMap.scala]: I wonder if we need to return 1000 results? Again, maybe not an issue but an increase in bandwidth by a factor of 10 (and we get a lot of requests for these resources).

**Comment by JustinPinner** [comment]
Looks good :+1:

File: applications/app/services/NewsSiteMap.scala (hunk 4 above)
Comment (rich-nguyen) [inline applications/app/services/NewsSiteMap.scala]: that's roughly how `largestEditorialCrop` behaves, picking a large non-4:3 crop, but falling back to largest image.

File: applications/app/services/NewsSiteMap.scala (hunk 5 above)
Comment (rich-nguyen) [inline applications/app/services/NewsSiteMap.scala]: I chose 1000 urls after speaking to chris M, who is unconvinced that these sitemaps have any positive or negative effect. The docs say 1000 urls (published within two days) is the news sitemap limit, so I'm hoping more urls will help news indexing...

https://support.google.com/news/publisher/answer/74288?hl=en-GB

File: applications/app/jobs/SitemapLifecycle.scala (hunk 1 above)
Comment (rich-nguyen) [inline applications/app/jobs/SitemapLifecycle.scala]: Good question, It's not clear what the effect would be of this. Maybe I should provide stale content (through the cdn) when the agent is cold.

Comment (rich-nguyen) [inline applications/app/jobs/SitemapLifecycle.scala]: done, sending 5xx to CDN to indicate stale should be served.

**Pull request merged**